samples/helpers/get-env/inc/get_env.hpp

```diff
@@ -0,0 +1,36 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+// Since `getenv()` may generate warnings on MSVC, and is not available on UWP, sample code
+// gets cluttered with insignificant nuances. This file makes it so that `std::getenv()` compiles
+// and works the same on Linux, macOS, Win32, and UWP.
+
+#pragma once
+
+#if !defined(_MSC_VER)
+
+// Linux and macOS
+#include <cstdlib>
+
+#else
+#define _CRT_SECURE_NO_WARNINGS
+
+#if !defined(WIN32_LEAN_AND_MEAN)
+#define WIN32_LEAN_AND_MEAN
+#endif
+#if !defined(NOMINMAX)
+#define NOMINMAX
+#endif
+#include <windows.h>
+
+#if !defined(WINAPI_PARTITION_DESKTOP) || WINAPI_PARTITION_DESKTOP
```

Comment:
why is this an OR ?   I see AND in the next file:

#if defined(WINAPI_PARTITION_DESKTOP) && !WINAPI_PARTITION_DESKTOP

Reply:
It is a negation, of that condition, look a `!`s.

Reply:
Basically, you always can transform `not ( A and not B )` into `not A or B`, i.e. to want to drop parenthesis which has `not` in front of it,, you apply negation to each term inside parenthesis, if you get double negation, you drop it, `and`s become `or`s and vice versa.

Reply:
true true , we learned that yesterday :)

Reply:
Not only that, but also the fact that `#if !defined(WINAPI_PARTITION_DESKTOP) || WINAPI_PARTITION_DESKTOP` means it is desktop, and otherwise it is UWP :) I know it looks confusing, but I didn't invent this macro.

At first, I got confused myself, and then I think I had to explain it to other people 3 or 4 times. Partially it is also confusing because visually it almost looks like `if DESKTOP or not DESKTOP`.

I think that can be explained by the fact that "partition" concept wasn't there initially, so wasn't defined the macro, and that's where `if not defined DESKTOP` comes from. And then, if defined, you can think of `DESKTOP` as boolean. The fact that it is defined does not yet mean that it is true.

So, if it was a C++ code, it would be something like a function with version 1 supporting Desktop only having a `f()`, version 2 having `enum Kind { Desktop, Phone, XBox }; f(Kind* kind = nullptr)`, so the code to check for isDesktop would be `bool isDesktop = (kind == nullptr || *kind == Kind::Desktop)`. Not exactly like that, because we'd rather have the default parameter value to be `Kind::Desktop`, but you get the idea.

I think what the API team should've done, is to have a macro defined that is called `UWP`, that is defined when NOT Desktop, and that's it. So, the condition for desktop would be simply `#ifndef UWP`, and then if UWP and you want to distinguish between phone and xbox, I'd make another variable called `UWP_KIND` or something. But I don't invent these.